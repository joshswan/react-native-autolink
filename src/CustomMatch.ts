import { Match, MatchConfig } from 'autolinker/dist/es2015';
import React from 'react';
import { StyleProp, TextStyle } from 'react-native';

// The variadic arguments of a regex replacer function, wrapped in an array.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ReplacerArgs = [string, ...any[]];

export interface CustomMatcher {
  /* Regular expression pattern to match/link user-specified patterns */
  pattern: RegExp;
  /* Custom press handler for links of this type */
  onPress?: (match: CustomMatch) => void;
  /* Custom long-press handler for links of this type */
  onLongPress?: (match: CustomMatch) => void;
  /* Custom styling for links of this type */
  style?: StyleProp<TextStyle>;
  /* Custom type/identifier for use with match.getType() */
  type?: string;
  /* Custom function for extracting link text using regex replacer args */
  getLinkText?: (replacerArgs: ReplacerArgs) => string;
  /* Custom function for extracting link URL using regex replacer args */
  getLinkUrl?: (replacerArgs: ReplacerArgs) => string;
  /* Custom function for rendering link - remember to attach press handlers! */
  renderLink?: (text: string, match: CustomMatch, index: number) => React.ReactNode;
}

export interface CustomMatchConfig extends MatchConfig {
  matcher: CustomMatcher;
  replacerArgs: ReplacerArgs;
}

export class CustomMatch extends Match {
  private matcher: CustomMatcher;
  private replacerArgs: ReplacerArgs;

  constructor({ matcher, replacerArgs, ...config }: CustomMatchConfig) {
    super(config);

    this.matcher = matcher;
    this.replacerArgs = replacerArgs;
  }

  getType(): string {
    return this.matcher.type || 'custom';
  }

  getAnchorHref(): string {
    return this.matcher.getLinkUrl?.(this.replacerArgs) ?? this.matchedText;
  }

  getAnchorText(): string {
    return this.matcher.getLinkText?.(this.replacerArgs) ?? this.matchedText;
  }

  getMatcher(): CustomMatcher {
    return this.matcher;
  }

  getReplacerArgs(): ReplacerArgs {
    return this.replacerArgs;
  }

  getRenderFn(): CustomMatcher['renderLink'] {
    return this.matcher.renderLink;
  }
}
