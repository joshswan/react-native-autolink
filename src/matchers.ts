/*!
 * React Native Autolink
 *
 * Copyright 2016-2020 Josh Swan
 * Released under the MIT license
 * https://github.com/joshswan/react-native-autolink/blob/master/LICENSE
 */

import { Match, MatchConfig } from 'autolinker/dist/es2015';

export interface LatLngMatchConfig extends MatchConfig {
  latlng: string;
}

export class LatLngMatch extends Match {
  private latlng: string;

  constructor(cfg: LatLngMatchConfig) {
    super(cfg);

    this.latlng = cfg.latlng;
  }

  getType(): string {
    return 'latlng';
  }

  getLatLng(): string {
    return this.latlng;
  }

  getAnchorHref(): string {
    return this.latlng;
  }

  getAnchorText(): string {
    return this.latlng;
  }
}

export const CustomMatchers = {
  // LatLng
  latlng: {
    id: 'latlng',
    regex: /[-+]?([1-8]?\d(\.\d+)?|90(\.0+)?),\s*[-+]?(180(\.0+)?|((1[0-7]\d)|([1-9]?\d))(\.\d+)?)/g,
    Match: LatLngMatch,
  },
};

export type MatcherId = keyof typeof CustomMatchers;

export const Matchers = Object.values(CustomMatchers);
