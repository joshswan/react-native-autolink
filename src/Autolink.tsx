/*!
 * React Native Autolink
 *
 * Copyright 2016-2021 Josh Swan
 * Released under the MIT license
 * https://github.com/joshswan/react-native-autolink/blob/master/LICENSE
 */

import React, { createElement, useCallback, useRef } from 'react';
import {
  Autolinker,
  AnchorTagBuilder,
  Match,
  EmailMatch,
  HashtagMatch,
  MentionMatch,
  PhoneMatch,
} from 'autolinker/dist/es2015';
import { Alert, Linking, StyleSheet, StyleProp, Text, TextStyle, TextProps } from 'react-native';
import { truncate } from './truncate';
import { CustomMatch, CustomMatcher } from './CustomMatch';
import { PolymorphicComponentProps } from './types';
import * as urls from './urls';

const makeTokenGenerator = (uid: string): [() => string, RegExp] => {
  let counter = 0;
  return [
    // eslint-disable-next-line no-plusplus
    () => `@__ELEMENT-${uid}-${counter++}__@`,
    new RegExp(`(@__ELEMENT-${uid}-\\d+__@)`, 'g'),
  ];
};

const styles = StyleSheet.create({
  link: {
    color: '#0E7AFE',
  },
});

const tagBuilder = new AnchorTagBuilder();

export interface AutolinkProps {
  email?: boolean;
  hashtag?: false | 'facebook' | 'instagram' | 'twitter';
  linkProps?: TextProps;
  linkStyle?: StyleProp<TextStyle>;
  matchers?: CustomMatcher[];
  mention?: false | 'instagram' | 'soundcloud' | 'twitter';
  onPress?: (url: string, match: Match) => void;
  onLongPress?: (url: string, match: Match) => void;
  phone?: boolean | 'text' | 'sms';
  renderLink?: (text: string, match: Match, index: number) => React.ReactNode;
  renderText?: (text: string, index: number) => React.ReactNode;
  showAlert?: boolean;
  stripPrefix?: boolean;
  stripTrailingSlash?: boolean;
  text: string;
  textProps?: TextProps;
  truncate?: number;
  truncateChars?: string;
  truncateLocation?: 'end' | 'middle' | 'smart';
  url?:
    | boolean
    | {
        schemeMatches?: boolean;
        wwwMatches?: boolean;
        tldMatches?: boolean;
      };
  useNativeSchemes?: boolean;
}

export type AutolinkComponentProps<C extends React.ElementType = typeof Text> = PolymorphicComponentProps<
  C,
  AutolinkProps
>;

export const Autolink = React.memo(
  <C extends React.ElementType = typeof Text>({
    as,
    component,
    email = true,
    hashtag = false,
    linkProps = {},
    linkStyle,
    matchers = [],
    mention = false,
    onPress: onPressProp,
    onLongPress: onLongPressProp,
    phone = false,
    renderLink: renderLinkProp,
    renderText,
    showAlert = false,
    stripPrefix = true,
    stripTrailingSlash = true,
    text,
    textProps = {},
    truncate: truncateProp = 0,
    truncateChars = '..',
    truncateLocation = 'smart',
    url = true,
    useNativeSchemes = false,
    ...props
  }: AutolinkComponentProps<C>): JSX.Element | null => {
    const getUrl = useCallback(
      (match: Match): string => {
        switch (match.getType()) {
          case 'email':
            return urls.getEmailUrl(match as EmailMatch);
          case 'hashtag':
            return urls.getHashtagUrl(match as HashtagMatch, hashtag, useNativeSchemes);
          case 'mention':
            return urls.getMentionUrl(match as MentionMatch, mention, useNativeSchemes);
          case 'phone':
            return urls.getPhoneUrl(match as PhoneMatch, phone);
          default:
            return match.getAnchorHref();
        }
      },
      [hashtag, mention, phone, useNativeSchemes],
    );

    const onPress = useCallback(
      (match: Match, alertShown?: boolean): void => {
        // Bypass default press handling if matcher has custom onPress
        if (match instanceof CustomMatch && match.getMatcher().onPress) {
          match.getMatcher().onPress?.(match);
          return;
        }

        // Check if alert needs to be shown
        if (showAlert && !alertShown) {
          Alert.alert('Leaving App', 'Do you want to continue?', [
            { text: 'Cancel', style: 'cancel' },
            { text: 'OK', onPress: () => onPress(match, true) },
          ]);
          return;
        }

        const linkUrl = getUrl(match);

        if (onPressProp) {
          onPressProp(linkUrl, match);
        } else {
          Linking.openURL(linkUrl);
        }
      },
      [getUrl, onPressProp, showAlert],
    );

    const onLongPress = useCallback(
      (match: Match): void => {
        // Bypass default press handling if matcher has custom onLongPress
        if (match instanceof CustomMatch && match.getMatcher().onLongPress) {
          match.getMatcher().onLongPress?.(match);
          return;
        }

        if (onLongPressProp) {
          const linkUrl = getUrl(match);
          onLongPressProp(linkUrl, match);
        }
      },
      [getUrl, onLongPressProp],
    );

    const renderLink = useCallback(
      (linkText: string, match: Match | CustomMatch, index: number) => {
        const truncated = truncateProp
          ? truncate(linkText, truncateProp, truncateChars, truncateLocation)
          : linkText;

        return (
          <Text
            style={
              (match instanceof CustomMatch && match.getMatcher().style) || linkStyle || styles.link
            }
            onPress={() => onPress(match)}
            onLongPress={() => onLongPress(match)}
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...linkProps}
            key={index}
          >
            {truncated}
          </Text>
        );
      },
      [linkProps, linkStyle, truncateProp, truncateChars, truncateLocation, onPress, onLongPress],
    );

    // Creates a token with a random UID that should not be guessable or
    // conflict with other parts of the string.
    const uid = useRef(Math.floor(Math.random() * 0x10000000000).toString(16));
    const [generateToken, tokenRegexp] = makeTokenGenerator(uid.current);

    const matches: { [token: string]: Match } = {};
    let linkedText: string;

    try {
      linkedText = Autolinker.link(text || '', {
        email,
        hashtag,
        mention,
        phone: !!phone,
        urls: url,
        stripPrefix,
        stripTrailingSlash,
        replaceFn: (match) => {
          const token = generateToken();

          matches[token] = match;

          return token;
        },
      });

      // User-specified custom matchers
      matchers.forEach((matcher) => {
        linkedText = linkedText.replace(matcher.pattern, (...replacerArgs) => {
          const token = generateToken();
          const matchedText = replacerArgs[0];

          matches[token] = new CustomMatch({
            matcher,
            matchedText,
            offset: replacerArgs[replacerArgs.length - 2],
            replacerArgs,
            tagBuilder,
          });

          return token;
        });
      });
    } catch (e) {
      // eslint-disable-next-line no-console
      console.warn('RN Autolink error:', e);
      return null;
    }

    const nodes = linkedText
      .split(tokenRegexp)
      .filter((part) => !!part)
      .map((part, index) => {
        const match = matches[part];

        // Check if rendering link or text node
        if (match?.getType()) {
          return (renderLinkProp || renderLink)(match.getAnchorText(), match, index);
        }

        return renderText ? (
          renderText(part, index)
        ) : (
          // eslint-disable-next-line react/jsx-props-no-spreading, react/no-array-index-key
          <Text {...textProps} key={index}>
            {part}
          </Text>
        );
      });

    return createElement(as || component || Text, props, ...nodes);
  },
);
