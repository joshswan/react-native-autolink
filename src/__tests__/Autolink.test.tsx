/*!
 * React Native Autolink
 *
 * Copyright 2016-2023 Josh Swan
 * Released under the MIT license
 * https://github.com/joshswan/react-native-autolink/blob/master/LICENSE
 */

import React from 'react';
import { Alert, Text, View } from 'react-native';
import renderer from 'react-test-renderer';
import { Autolink } from '../Autolink';
import { CustomMatch } from '../CustomMatch';
import { IntlPhoneMatcher, LatLngMatcher } from '../matchers';

describe('<Autolink />', () => {
  test('renders a Text node', () => {
    const tree = renderer.create(<Autolink text="" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders a string when nothing to link', () => {
    const tree = renderer.create(<Autolink text="Testing" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders a custom container node', () => {
    const tree = renderer.create(<Autolink component={View} text="Testing" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('wraps an email address with a link Text node when email prop enabled', () => {
    const tree = renderer.create(<Autolink text="josh@example.com" email />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('does not wrap an email address in a link Text node when email prop disabled', () => {
    const tree = renderer.create(<Autolink text="josh@example.com" email={false} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('wraps a hashtag in a link Text node when hashtag prop enabled', () => {
    const tree = renderer.create(<Autolink text="#awesome" hashtag="instagram" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('does not wrap a hashtag in a link Text node when hashtag prop disabled', () => {
    const tree = renderer.create(<Autolink text="#awesome" hashtag={false} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('wraps a mention/handle in a link Text node when mention prop enabled', () => {
    const tree = renderer.create(<Autolink text="@twitter" mention="twitter" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('does not wrap a mention/handle in a link Text node when mention prop disabled', () => {
    const tree = renderer.create(<Autolink text="@twitter" mention={false} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('wraps a phone number in a link Text node when phone prop enabled', () => {
    const tree = renderer.create(<Autolink text="415-555-5555" phone />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('does not wrap a phone number in a link Text node when phone prop disabled', () => {
    const tree = renderer.create(<Autolink text="415-555-5555" phone={false} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('wraps a url in a link Text node when url prop enabled', () => {
    const tree = renderer
      .create(<Autolink text="https://github.com/joshswan/react-native-autolink" url />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('matches top-level domain url when wwwMatches enabled', () => {
    const tree = renderer
      .create(<Autolink text="github.com" url={{ tldMatches: true }} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('does not match top-level domain url when wwwMatches disabled', () => {
    const tree = renderer
      .create(<Autolink text="github.com" url={{ tldMatches: false }} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('does not match www-containing url when wwwMatches disabled', () => {
    const tree = renderer
      .create(<Autolink text="www.github.com" url={{ wwwMatches: false }} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('does not match scheme-containing url when schemeMatches disabled', () => {
    const tree = renderer
      .create(<Autolink text="http://github.com" url={{ schemeMatches: false }} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('does not wrap a url in a link Text node when url prop disabled', () => {
    const tree = renderer
      .create(<Autolink text="https://github.com/joshswan/react-native-autolink" url={false} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('links multiple elements individually', () => {
    const tree = renderer
      .create(
        <Autolink
          text="Hi @josh (josh@example.com or 415-555-5555), check out https://github.com/joshswan/react-native-autolink. It's #awesome!"
          email
          hashtag="instagram"
          mention="twitter"
          phone
          url
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('removes url prefixes when stripPrefix prop enabled', () => {
    const tree = renderer.create(<Autolink text="https://github.com" stripPrefix />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('does not remove url prefixes when stripPrefix prop disabled', () => {
    const tree = renderer
      .create(<Autolink text="https://github.com" stripPrefix={false} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('removes url trailing slashes when stripTrailingSlash prop enabled', () => {
    const tree = renderer.create(<Autolink text="github.com/" stripTrailingSlash />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('does not remove url trailing slashes when stripTrailingSlash prop disabled', () => {
    const tree = renderer
      .create(<Autolink text="github.com/joshswan/" stripTrailingSlash={false} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('truncates urls to length specified in truncate prop', () => {
    const tree = renderer
      .create(<Autolink text="github.com/joshswan/react-native-autolink" truncate={32} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('replaces removed protion of truncated url with truncateChars prop value', () => {
    const tree = renderer
      .create(
        <Autolink
          text="github.com/joshswan/react-native-autolink"
          truncate={32}
          truncateChars="__"
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('truncates urls at the end when specified in truncateLocation prop', () => {
    const tree = renderer
      .create(
        <Autolink
          stripPrefix={false}
          text="https://github.com/joshswan/react-native-autolink"
          truncate={32}
          truncateLocation="end"
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('truncates urls in the middle when specified in truncateLocation prop', () => {
    const tree = renderer
      .create(
        <Autolink
          stripPrefix={false}
          text="https://github.com/joshswan/react-native-autolink"
          truncate={32}
          truncateLocation="middle"
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders links using renderLink prop if provided', () => {
    const renderLink = (text, match, index) => <Text>{`${text}:${index}`}</Text>;
    const tree = renderer
      .create(<Autolink text="josh@example.com" renderLink={renderLink} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders text using renderText prop if provided', () => {
    const renderText = (text) => (
      <View>
        <Text>{text}</Text>
      </View>
    );
    const tree = renderer
      .create(<Autolink component={View} text="Testing" renderText={renderText} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('calls onPress handler prop when link clicked', () => {
    const onPress = jest.fn();
    const tree = renderer.create(<Autolink text="josh@example.com" onPress={onPress} />);
    tree.root.findAllByType(Text)[1].props.onPress();
    expect(onPress.mock.calls.length).toBe(1);
    expect(onPress.mock.calls[0][0]).toBe('mailto:josh%40example.com');
  });

  test('calls onLongPress handler prop when link long pressed', () => {
    const onLongPress = jest.fn();
    const tree = renderer.create(<Autolink text="josh@example.com" onLongPress={onLongPress} />);
    tree.root.findAllByType(Text)[1].props.onLongPress();
    expect(onLongPress.mock.calls.length).toBe(1);
    expect(onLongPress.mock.calls[0][0]).toBe('mailto:josh%40example.com');
  });

  describe('alert', () => {
    test('displays alert before linking when showAlert prop enabled', () => {
      const spy = jest.spyOn(Alert, 'alert').mockImplementationOnce(() => {});
      const tree = renderer.create(<Autolink text="#awesome" hashtag="instagram" showAlert />);
      tree.root.findAllByType(Text)[1].props.onPress();
      expect(spy).toBeCalledTimes(1);
      expect(spy).toBeCalledWith('Leaving App', 'Do you want to continue?', expect.any(Array));
    });
  });

  describe('urls', () => {
    test('uses hashtag url when pressing linked hashtag', () => {
      const onPress = jest.fn();
      const tree = renderer.create(
        <Autolink text="#awesome" hashtag="instagram" onPress={onPress} />,
      );
      tree.root.findAllByType(Text)[1].props.onPress();
      expect(onPress.mock.calls.length).toBe(1);
      expect(onPress.mock.calls[0][0]).toBe('https://www.instagram.com/explore/tags/awesome/');
    });

    test('uses mention url when pressing linked mention', () => {
      const onPress = jest.fn();
      const tree = renderer.create(
        <Autolink text="@twitter" mention="twitter" onPress={onPress} />,
      );
      tree.root.findAllByType(Text)[1].props.onPress();
      expect(onPress.mock.calls.length).toBe(1);
      expect(onPress.mock.calls[0][0]).toBe('https://twitter.com/twitter');
    });

    test('uses native scheme for mention url when enabled', () => {
      const onPress = jest.fn();
      const tree = renderer.create(
        <Autolink text="@twitter" mention="twitter" onPress={onPress} useNativeSchemes />,
      );
      tree.root.findAllByType(Text)[1].props.onPress();
      expect(onPress.mock.calls.length).toBe(1);
      expect(onPress.mock.calls[0][0]).toBe('twitter://user?screen_name=twitter');
    });

    test('uses phone url when pressing linked phone number', () => {
      const onPress = jest.fn();
      const tree = renderer.create(<Autolink text="415-555-5555" phone onPress={onPress} />);
      tree.root.findAllByType(Text)[1].props.onPress();
      expect(onPress.mock.calls.length).toBe(1);
      expect(onPress.mock.calls[0][0]).toBe('tel:4155555555');
    });
  });

  /**
   * Custom matchers
   */
  describe('matchers', () => {
    test('wraps text based on supplied custom matchers', () => {
      const tree = renderer
        .create(<Autolink text="34.0522, -118.2437" matchers={[LatLngMatcher]} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('calls custom onPress handlers', () => {
      const onPress = jest.fn();
      const tree = renderer.create(
        <Autolink text="+14085550123" matchers={[{ ...IntlPhoneMatcher, onPress }]} />,
      );
      tree.root.findAllByType(Text)[1].props.onPress();
      expect(onPress.mock.calls.length).toBe(1);
      expect(onPress.mock.calls[0][0]).toBeInstanceOf(CustomMatch);
    });

    test('calls custom onLongPress handlers', () => {
      const onLongPress = jest.fn();
      const tree = renderer.create(
        <Autolink text="+14085550123" matchers={[{ ...IntlPhoneMatcher, onLongPress }]} />,
      );
      tree.root.findAllByType(Text)[1].props.onLongPress();
      expect(onLongPress.mock.calls.length).toBe(1);
      expect(onLongPress.mock.calls[0][0]).toBeInstanceOf(CustomMatch);
    });

    test('uses getLinkText when rendering link', () => {
      const tree = renderer
        .create(
          <Autolink
            text="+14085550123"
            matchers={[{ ...IntlPhoneMatcher, getLinkText: () => '__LINK_TEXT__' }]}
          />,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('uses getLinkUrl when using default onPress handler', () => {
      const onPress = jest.fn();
      const tree = renderer.create(
        <Autolink
          text="+14085550123"
          onPress={onPress}
          matchers={[{ ...IntlPhoneMatcher, getLinkUrl: () => '__LINK_URL__' }]}
        />,
      );
      tree.root.findAllByType(Text)[1].props.onPress();
      expect(onPress.mock.calls.length).toBe(1);
      expect(onPress.mock.calls[0][0]).toBe('__LINK_URL__');
      expect(onPress.mock.calls[0][1]).toBeInstanceOf(CustomMatch);
    });

    test('uses renderLink when rendering link', () => {
      const onPress = jest.fn();
      const renderLink = jest
        .fn()
        .mockImplementation((text: string) => <Text onPress={onPress}>{text}</Text>);
      const tree = renderer.create(
        <Autolink
          text="+14085550123"
          onPress={onPress}
          matchers={[{ ...IntlPhoneMatcher, renderLink }]}
        />,
      );
      expect(renderLink.mock.calls.length).toBe(1);
      expect(renderLink.mock.calls[0][0]).toBe('+14085550123');
      expect(renderLink.mock.calls[0][1]).toBeInstanceOf(CustomMatch);
      expect(renderLink.mock.calls[0][2]).toBe(0);
      expect(tree).toMatchSnapshot();
      tree.root.findAllByType(Text)[1].props.onPress();
      expect(onPress.mock.calls.length).toBe(1);
    });
  });
});
