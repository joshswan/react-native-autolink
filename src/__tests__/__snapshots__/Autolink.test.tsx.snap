// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Autolink /> does not match scheme-containing url when schemeMatches disabled 1`] = `
<Text>
  <Text>
    http://github.com
  </Text>
</Text>
`;

exports[`<Autolink /> does not match top-level domain url when wwwMatches disabled 1`] = `
<Text>
  <Text>
    github.com
  </Text>
</Text>
`;

exports[`<Autolink /> does not match www-containing url when wwwMatches disabled 1`] = `
<Text>
  <Text>
    www.github.com
  </Text>
</Text>
`;

exports[`<Autolink /> does not remove url prefixes when stripPrefix prop disabled 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    https://github.com
  </Text>
</Text>
`;

exports[`<Autolink /> does not remove url trailing slashes when stripTrailingSlash prop disabled 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    github.com/joshswan/
  </Text>
</Text>
`;

exports[`<Autolink /> does not wrap a hashtag in a link Text node when hashtag prop disabled 1`] = `
<Text>
  <Text>
    #awesome
  </Text>
</Text>
`;

exports[`<Autolink /> does not wrap a mention/handle in a link Text node when mention prop disabled 1`] = `
<Text>
  <Text>
    @twitter
  </Text>
</Text>
`;

exports[`<Autolink /> does not wrap a phone number in a link Text node when phone prop disabled 1`] = `
<Text>
  <Text>
    415-555-5555
  </Text>
</Text>
`;

exports[`<Autolink /> does not wrap a url in a link Text node when url prop disabled 1`] = `
<Text>
  <Text>
    https://github.com/joshswan/react-native-autolink
  </Text>
</Text>
`;

exports[`<Autolink /> does not wrap an email address in a link Text node when email prop disabled 1`] = `
<Text>
  <Text>
    josh@example.com
  </Text>
</Text>
`;

exports[`<Autolink /> links multiple elements individually 1`] = `
<Text>
  <Text>
    Hi 
  </Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    @josh
  </Text>
  <Text>
     (
  </Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    josh@example.com
  </Text>
  <Text>
     or 
  </Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    415-555-5555
  </Text>
  <Text>
    ), check out 
  </Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    github.com/joshswan/react-native-autolink
  </Text>
  <Text>
    . It's 
  </Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    #awesome
  </Text>
  <Text>
    !
  </Text>
</Text>
`;

exports[`<Autolink /> matchers uses getLinkText when rendering link 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    __LINK_TEXT__
  </Text>
</Text>
`;

exports[`<Autolink /> matchers uses renderLink when rendering link 1`] = `
<Text>
  <Text
    onPress={[MockFunction]}
  >
    +14085550123
  </Text>
</Text>
`;

exports[`<Autolink /> matchers wraps text based on supplied custom matchers 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    34.0522, -118.2437
  </Text>
</Text>
`;

exports[`<Autolink /> matches top-level domain url when wwwMatches enabled 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    github.com
  </Text>
</Text>
`;

exports[`<Autolink /> removes url prefixes when stripPrefix prop enabled 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    github.com
  </Text>
</Text>
`;

exports[`<Autolink /> removes url trailing slashes when stripTrailingSlash prop enabled 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    github.com
  </Text>
</Text>
`;

exports[`<Autolink /> renders a Text node 1`] = `<Text />`;

exports[`<Autolink /> renders a custom container node 1`] = `
<View>
  <Text>
    Testing
  </Text>
</View>
`;

exports[`<Autolink /> renders a string when nothing to link 1`] = `
<Text>
  <Text>
    Testing
  </Text>
</Text>
`;

exports[`<Autolink /> renders links using renderLink prop if provided 1`] = `
<Text>
  <Text>
    josh@example.com:0
  </Text>
</Text>
`;

exports[`<Autolink /> renders text using renderText prop if provided 1`] = `
<View>
  <View>
    <Text>
      Testing
    </Text>
  </View>
</View>
`;

exports[`<Autolink /> replaces removed protion of truncated url with truncateChars prop value 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    github.com/joshswan/__e-autolink
  </Text>
</Text>
`;

exports[`<Autolink /> truncates urls at the end when specified in truncateLocation prop 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    https://github.com/joshswan/re..
  </Text>
</Text>
`;

exports[`<Autolink /> truncates urls in the middle when specified in truncateLocation prop 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    https://github...native-autolink
  </Text>
</Text>
`;

exports[`<Autolink /> truncates urls to length specified in truncate prop 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    github.com/joshswan/..e-autolink
  </Text>
</Text>
`;

exports[`<Autolink /> wraps a hashtag in a link Text node when hashtag prop enabled 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    #awesome
  </Text>
</Text>
`;

exports[`<Autolink /> wraps a mention/handle in a link Text node when mention prop enabled 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    @twitter
  </Text>
</Text>
`;

exports[`<Autolink /> wraps a phone number in a link Text node when phone prop enabled 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    415-555-5555
  </Text>
</Text>
`;

exports[`<Autolink /> wraps a url in a link Text node when url prop enabled 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    github.com/joshswan/react-native-autolink
  </Text>
</Text>
`;

exports[`<Autolink /> wraps an email address with a link Text node when email prop enabled 1`] = `
<Text>
  <Text
    onLongPress={[Function]}
    onPress={[Function]}
    style={
      {
        "color": "#0E7AFE",
      }
    }
  >
    josh@example.com
  </Text>
</Text>
`;
